### ISSUE TITLE
Switch away from conda

### ISSUE THREAD

**Issue by leoisl** [mbhall88/tbpore#34]
This is a request from Floriane and Pedro from our meeting on 18/05/2022.

The cluster they will use to run `tbpore` in Canada has a policy to not allow users to run `conda` because it messes up with the `~/.bashrc` configuration. Is there an alternative way for us to provide the dependencies? `pypi` would be great, but IDK if all dependencies are installable through `pypi`, my guess is not e.g. `psdm`, `rasusa`, etc... @mbhall88 do you have any suggestions? They said that they managed to run `tbpore` for now, by installing the dependencies through `conda`, but IIRC in a virtual machine (or a container?) in their cluster, but they also said this is not a long-term solution...

**Comment by mbhall88** [comment]
Ooof that's a bummer.

Making the whole thing available through pypi would mean reimplementing quite a few steps in python, which would take a fair bit of time.

Is a container acceptable to them? 

Alternatively, I think most (all?) of the dependencies are downloadable as binaries. Actually, mykrobe isn't. Not sure about samtools/bcftools - although they're quite easy to install.

So if container is not an option, we might need a Justfile/Makefile recipe to install/download the dependencies 🙁

**Comment by leoisl** [comment]
I will ask them about containers. Hope these will be fine, and then we can still make the conda recipe as it will give us a biocontainer anyway. Also if we reduce the RAM usage or if it is already runnable on Madagascar local machines, some of the samples might be analysed directly there and a conda recipe might be useful...

**Comment by pedrodcb** [comment]
Hello both,

Following up on this now, a container would be an option since Compute Canada makes using Singularity fairly easy. See: https://docs.alliancecan.ca/wiki/Singularity

Would that be an option? If not perhaps I could download the dependencies either as binary or in a python environment (e.g.: virtualenv).

Thank you!

**Comment by mbhall88** [comment]
Hi @pedrodcb 

There is a container that gets automatically built everytime a new conda build completes. However, it seems to be missing at the moment. Once https://github.com/bioconda/bioconda-recipes/issues/35824 is fixed, you can find the instructions for using that container [here](https://github.com/mbhall88/tbpore#container).

**Comment by mbhall88** [comment]
It's now available

**Comment by leoisl** [comment]
Hello @pedrodcb , did the bioconda singularity container work for you?

**Comment by pedrodcb** [comment]
Hello @leoisl, pardon the late reply. We've tried using the image but upon running it we stumble on the following error:

Command:
`singularity exec --cleanenv --no-home --bind $PWD:$PWD "$URI" tbpore process -o sample_example/tbpore_out --cleanup sample_example/mada_1-7.subsampled.fastq.gz`

Output:
```
INFO:    Using cached SIF image
[2022-10-04 15:32:21] INFO     | Welcome to TBpore version 0.1.0
Traceback (most recent call last):
  File "/usr/local/bin/tbpore", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.9/site-packages/tbpore/tbpore.py", line 473, in main
    main_cli()
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.9/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/tbpore/tbpore.py", line 215, in process
    config = load_config_file()
  File "/usr/local/lib/python3.9/site-packages/tbpore/tbpore.py", line 38, in load_config_file
    with open(config_file, "r") as config_file_fh:
FileNotFoundError: [Errno 2] No such file or directory: '/usr/local/lib/python3.9/site-packages/.config.yaml'
```

Perhaps the .config.yaml file is missing from the image? We've tried binding it but the image seems to be read-only. Anything you could do about this? We could still go the PyPI way but that would be a bit complicated, getting the image to work could save us some headache.

**Comment by mbhall88** [comment]
Ah, yes, it looks like this file (and a few others) were mistakenly not included in the distribution. I have fixed this with version 0.1.1. A new container should be up in the next few hours

**Comment by pedrodcb** [comment]
Hello @mbhall88, thank you very much for this! I've seen the update you've pushed but I can't seem to find the new container, could you point me to it? Should it be here?: https://quay.io/repository/biocontainers/tbpore?tab=tags

Thanks again!

**Comment by leoisl** [comment]
Hello @pedrodcb . We are waiting for the `conda` team to approve the [recipe](https://github.com/bioconda/bioconda-recipes/pull/37293) that @mbhall88 updated, once is approved and merged a biocontainer will be automatically created... but this might take some more days...

**Comment by mbhall88** [comment]
Thanks a lot @rpetit3. 

@pedrodcb let me know if you have any issues with that container.

**Comment by pedrodcb** [comment]
Thank you all for your help and for uploading the new container, it solved the previous issue but, @mbhall88, I'm now getting this error:
`ERROR    | Decontamination DB index /usr/local/lib/python3.9/site-packages/data/decontamination_db/tbpore.remove_contam.fa.gz.map-ont.mmi does not exist, please follow the instructions at https://github.com/mbhall88/tbpore#configuring-the-decontamination-database-index to download and configure it before running tbpore`

I've followed the instructions to configure the decontamination database properly and since I think that file would be a bit too big to fit in the container would you know how I could solve this?

The downloaded file is in our main work folder as described in the documentation:
`data/decontamination_db/tbpore.remove_contam.fa.gz.map-ont.mmi`

And this is the command I'm currently using to run the container:
`singularity exec -H /path_to_working_dir/tbpore tbpore_0.1.1.sif tbpore process -o sample_example/tbpore_out --cleanup sample_example/mada_1-7.subsampled.fastq.gz`

The reason we're not using the "$URI" is that building the container was not working on Compute Canada's Narval cluster for some reason, so we built it locally instead (`singularity build tbpore_0.1.1.sif docker://quay.io/biocontainers/tbpore:0.1.1--pyhdfd78af_0`) and then transferred it to be used on the cluster.

Thanks again!

**Comment by mbhall88** [comment]
Sorry about this @pedrodcb. Once #42 is merged, I'll release v0.2.0 which will allow to specify a path other than the default.

**Comment by mbhall88** [comment]
@pedrodcb v0.2.0 is released and there should be a container for it also. It allows you to specify a non-default location for the database (`--db`). It also updates mykrobe to the latest version which includes the new catalogue
